_posts/2020_04_29_encrypted_keras_with_syft/encrypted_keras_with_syft.Rmd

```diff
@@ -0,0 +1,413 @@
+  gregs <- runif(1, min = -Q, max = Q)
+  mine <- (x - julias - gregs) %% Q
+  list (julias, gregs, mine)
+}
+
+# some top secret value no-one may get to see
+value <- 77777
+
+encrypted <- encrypt(value)
+encrypted
+```
+
+    [[1]]
+    [1] 7467283737857
+
+    [[2]]
+    [1] 36307804406429
+
+    [[3]]
+    [1] 34315485297318
+
+Once the three of us put our shares together, getting back the plain value is straightforward:
+
+```{r}
+decrypt <- function(shares) {
+  Reduce(sum, shares) %% Q  
+}
+
+decrypt(encrypted)
+```
+
+    77777
+
+As an example of how to compute on encrypted data, here's addition. (Other operations will be a lot less straightforward.) To
+add two numbers, just have everyone add their respective shares:
+
+```{r}
+add <- function(x, y) {
+  list(
+    # julia
+    (x[[1]] + y[[1]]) %% Q,
+    # greg
+    (x[[2]] + y[[2]]) %% Q,
+    # me
+    (x[[3]] + y[[3]]) %% Q
+  )
+}
+  
+x <- encrypt(11)
+y <- encrypt(122)
+
+decrypt(add(x, y))
+```
+
+    133
+
+Back to the setting of deep learning and the current task to be solved: Have the server apply gradient updates without ever
+seeing them. With secret sharing, it would work like this:
+
+Julia, Greg and me each want to train on our own private data. Together, we will be responsible for gradient averaging, that
+is, we'll form a *cluster* of *workers* united in that task. Now, the model owner *secret shares* the model, and we start
+training, each on their own data. After some number of iterations, we use secure averaging to combine our respective
+gradients. Then, all the server gets to see is the mean gradient, and there is no way to determine our respective
+contributions.
+
+#### Beyond private gradients
+
+Amazingly, it is even possible to *train* on encrypted data -- amongst others, using that same technique of secret sharing. Of
+course, this has to negatively affect training speed. But it's good to know that if one's use case were to demand it, it would
+be feasible. (One possible use case is when training on one party's data alone doesn't make any sense, but data is sensitive,
+so others won't let you access their data unless encrypted.)
+
+So with encryption available on an all-you-need basis, are we completely safe, privacy-wise? The answer is no. The model can
+still leak information. For example, in some cases it is possible to perform *model inversion* [@abs-1805-04049], that is,
+with just black-box access to a model, train an *attack model* that allows reconstructing some of the original training data.
+Needless to say, this kind of leakage has to be avoided. [Differential
+privacy](https://blogs.rstudio.com/tensorflow/posts/2019-12-20-differential-privacy/) [@Dwork2006], [@dwork2006differential]
+demands that results obtained from querying a model be independent from the presence or absence, in the dataset employed for
+training, of a single individual. In general, this is ensured by adding noise to the answer to every query. In training deep
```

Comment:
Excellent, clear definition. I'd try to somehow emphasize definitions where the occur so that they're easier to find.